README
NAME
    `Devel::MAT' - Perl Memory Analysis Tool

DESCRIPTION
    A `Devel::MAT' instance loads a heapdump file, and provides a container
    to store analysis tools to work on it. Tools may be provided that
    conform to the Devel::MAT::Tool API, which can help analyse the data and
    interact with the explorer user interface by using the methods in the
    Devel::MAT::UI package.

  File Format
    The dump file format is still under development, so at present no
    guarantees are made on whether files can be loaded over mismatching
    versions of `Devel::MAT'. However, as of version 0.11 the format should
    be more extensible, allowing new SV fields to be added without breaking
    loading - older tools will ignore new fields and newer tools will just
    load undef for fields absent in older files. As the distribution
    approaches maturity the format will be made more stable.

CONSTRUCTOR
  $pmat = Devel::MAT->load( $path, %args )
    Loads a heap dump file from the given path, and returns a new
    `Devel::MAT' instance wrapping it.

METHODS
  $df = $pmat->dumpfile
    Returns the underlying Devel::MAT::Dumpfile instance backing this
    analysis object.

  @tools = $pmat->available_tools
    Lists the Devel::MAT::Tool classes that are installed and available.

  $tool = $pmat->load_tool( $name )
    Loads the named Devel::MAT::Tool class.

  $node = $pmat->inref_graph( $sv, %opts )
    Traces the tree of inrefs from `$sv' back towards the known roots,
    returning a Devel::MAT::Graph node object representing it, within a
    graph of reverse references back to the known roots.

    This method will load Devel::MAT::Tool::Inrefs if it isn't yet loaded.

    The following named options are recognised:

    depth => INT
        If specified, stop recursing after the specified count. A depth of 1
        will only include immediately referring SVs, 2 will print the
        referrers of those, etc. Nodes with inrefs that were trimmed because
        of this limit will appear to be roots with a special name of
        `EDEPTH'.

    strong => BOOL
    direct => BOOL
        Specifies the type of inrefs followed. By default all inrefs are
        followed. Passing `strong' will follow only strong direct inrefs.
        Passing `direct' will follow only direct inrefs.

  $sv = $pmat->find_symbol( $name )
    Attempts to walk the symbol table looking for a symbol of the given
    name, which must include the sigil.

     $Package::Name::symbol_name => to return a SCALAR SV
     @Package::Name::symbol_name => to return an ARRAY SV
     %Package::Name::symbol_name => to return a HASH SV
     &Package::Name::symbol_name => to return a CODE SV

  $gv = $pmat->find_glob( $name )
    Attempts to walk to the symbol table looking for a symbol of the given
    name, returning the `GLOB' object if found.

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Devel-MAT

0.20    2014/12/04 18:35:09
        [CHANGES]
         * Allow serial-numbering of .pmat files in the Dumper
         * Totally refactored $pmat->identify to return an abstract Graph
           structure
         * Much improved pmat-identify-sv output; sort strong-refs first,
           use Unicode linedrawing, identify "already seen" SVs clearer
         * GTK Explorer improvements:
            + Allow search by symbol
            + Show longer PV strings in tooltip
            + Selectable SV detail labels

0.19    2014/11/18 13:17:28
        [CHANGES]
         * No need to annotate 'directly' or 'indirectly' on outref names
         * Added -dump_at_DIE option to Devel::MAT::Dumper
         * Added icon for INVLIST from perl 5.20

0.18    2014/07/14 22:45:08
        [CHANGES]
         * Improved for Tools to create custom UIs
            + Widget trees
            + Label text + icon
            + Tool-localised icon load paths
         * Support loading Tools automatically
         * Neater API for inrefs/outrefs that also indicates link strength

0.17    2014/01/22 21:33:45
        [CHANGES]
         * Also dump SvUTF8() flag of PVs
         * Added pmat-cat-svpv

        [BUGFIXES]
         * Don't crash on NULL PADLIST elements (RT92290)

0.16    2014/01/15 18:18:53
        [CHANGES]
         * More accurate format for dumping SV MAGIC annotations
            - note that older files will contain incorrect information
         * More accurate backref information
         * Improvements to analysis of PAD lexicals

        [BUGFIXES]
         * Fix many cases of optree dumping to get correct PAD indexes
         * Further minor fixes that improve SvREFCNT / inrefs accuracy

0.15    2014/01/14 11:12:37
        [CHANGES]
         * Remember to basename() $0 for default dumpfile name
         * Added -eager_option import option
         * Added $df->roots_strong
         * Added options to $pmat->identify and pmat-identify-sv command

        [BUGFIXES]
         * Avoid infinite loop in $pmat->identify in cyclic cases
         * Fix docs on ->rv method (RT92122)

0.14    2014/01/04 20:25:16
        [CHANGES]
         * Move much core data storage into XS structs, for reduced memory
           consumption of analyser and some CPU performance gains
         * Display human-readable size information in GTK explorer
         * Added a size totals summary commandline tool
         * Improved recursion logic in Inrefs tool; avoids multiple recursions
           into shared structures

        [BUGFIXES]
         * Fix for 5.10.0

0.13    2013/12/06 21:37:35
        [CHANGES]
         * Performance improvement to inrefs counting
         * Added Sizes tool - calculates structure and owned set and sizes
         * Allow Tools to declare UI radiobutton sets
         * Improved Tools ability to interact with the SV list
         * Disable Tools menu items for loaded tools to avoid loading them
           twice

        [BUGFIXES]
         * Add List::UtilsBy to requirements
         * Our version of dopoptosub_at() doesn't need aTHX_

0.12    2013/12/03 15:18:43
        [CHANGES]
         * Classify outrefs and inrefs into various kinds
         * Subclass ARRAY into PADLIST, PADNAMES, PAD for CODE
         * Allow selection of what kinds of refs to count/display on the GTK
           explorer
         * Store more flags (CvWEAKOUTSIDE, CvCVGV_RC, AvREAL) so as to get
           strong/weak references correct
         * Store the SvOUTSTASH of SCALARs and REFs
         * Added a commandline wrapper for $pmat->identify

        [BUGFIXES]
         * require() tool .pm files automatically when calling ->load_tool

0.11    2013/11/28 14:30:32
        [CHANGES]
         * Complete redesign of dumpfile format so it is more extensible in future
         * Represent RVs by their own type, REF, rather than being a kind of SCALAR
         * Store caller()-like context in dumpfiles as well
         * Store CLONE/CLONED/ISXSUB CV flags and oproot
         * Attempt to match CLONE protosubs with CLONED closures
         * Added SCALAR->iv method
         * Neaten the names of STASHes whose names begin with a control character
         * Pretty-print PVs and limit length to 32 characters in GTK explorer
         * Added icons for REF and LVALUE types

        [BUGFIXES]
         * Guard against PL_main_root being NULL, as it can be early in compilation
         * Improved Reachability tool logic for CODE PAD slots

0.10    2013/11/26 00:42:14
        [CHANGES]
         * Display dumped perl version, SV and byte counts in GTK explorer
           status bar
         * Dump all GVs, even those without GPs
         * Store size information about all SVs
            + TODO: CODE size does not take account of optree
         * Have dump_at_{END,SIGABRT,SIGQUIT} print to STDERR
         * Limit the capture size of PVs, but still note the full length
         * Dump FILE and LINE of CVs and GVs
         * Redrawn type icons

        [BUGFIXES]
         * Fix for NVLEN=16 architectures

0.09    2013/11/22 00:25:17
        [CHANGES]
         * Remove orphan count unit test as it is too unreliable

        [BUGFIXES]
         * Sometimes a PAD(0) slot is NULL
         * ->identify is now a $pmat method
         * ->identify should load Inrefs tool
         * Try harder to clear PL_tmpsv when making an unreachable cycle

0.08    2013/11/10 19:27:30
        [CHANGES]
         * Chase more fields of an HV's mro_meta
         * Document how Tools may interact with the UI
         * Allow Tools to register and use icons on the UI
         * Rework Reachability tool to analyse kinds of reachability
         * Move inrefs logic into its own new Inrefs Tool
         * Slight performance improvement to dumpfile loading
         * Display icons also for general SV type

        [BUGFIXES]
         * Ensure ->find_symbol actually works for deeply-nested symbols

0.07    2013/11/08 01:33:01

lib/Devel/MAT.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2013-2014 -- leonerd@leonerd.org.uk

package Devel::MAT;

use strict;
use warnings;

our $VERSION = '0.20';

use Carp;
use List::Util qw( pairs );
use List::UtilsBy qw( sort_by );

use Devel::MAT::Dumpfile;
use Devel::MAT::Graph;

use Module::Pluggable
   sub_name => "_available_tools",
   search_path => [ "Devel::MAT::Tool" ],
   require => 1;

require XSLoader;
XSLoader::load( __PACKAGE__, $VERSION );

=head1 NAME

C<Devel::MAT> - Perl Memory Analysis Tool

=head1 DESCRIPTION

A C<Devel::MAT> instance loads a heapdump file, and provides a container to
store analysis tools to work on it. Tools may be provided that conform to the
L<Devel::MAT::Tool> API, which can help analyse the data and interact with the
explorer user interface by using the methods in the L<Devel::MAT::UI> package.

=head2 File Format

The dump file format is still under development, so at present no guarantees
are made on whether files can be loaded over mismatching versions of
C<Devel::MAT>. However, as of version 0.11 the format should be more
extensible, allowing new SV fields to be added without breaking loading - older
tools will ignore new fields and newer tools will just load undef for fields
absent in older files. As the distribution approaches maturity the format will
be made more stable.

=cut

=head1 CONSTRUCTOR

=cut

=head2 $pmat = Devel::MAT->load( $path, %args )

Loads a heap dump file from the given path, and returns a new C<Devel::MAT>
instance wrapping it.

=cut

sub load
{
   my $class = shift;

   my $df = Devel::MAT::Dumpfile->load( @_ );

   return bless {
      df => $df,
   }, $class;
}

=head1 METHODS

=cut

=head2 $df = $pmat->dumpfile

Returns the underlying L<Devel::MAT::Dumpfile> instance backing this analysis
object.

=cut

sub dumpfile
{
   my $self = shift;
   return $self->{df};
}

=head2 @tools = $pmat->available_tools

Lists the L<Devel::MAT::Tool> classes that are installed and available.

=cut

{
   my @TOOLS;
   my $TOOLS_LOADED;

   sub available_tools
   {
      my $self = shift;

      return @TOOLS if $TOOLS_LOADED;

      $TOOLS_LOADED++;
      @TOOLS = map { $_ =~ s/^Devel::MAT::Tool:://; $_ } $self->_available_tools;

      foreach my $name ( @TOOLS ) {
         my $tool_class = "Devel::MAT::Tool::$name";
         next unless $tool_class->can( "AUTOLOAD_TOOL" ) and $tool_class->AUTOLOAD_TOOL( $self );

         $self->{tools}{$name} ||= $tool_class->new( $self );
      }

      return @TOOLS;
   }
}

=head2 $tool = $pmat->load_tool( $name )

Loads the named L<Devel::MAT::Tool> class.

=cut

sub load_tool
{
   my $self = shift;
   my ( $name, %args ) = @_;

   # Ensure tools are 'require'd
   $self->available_tools;

   my $tool_class = "Devel::MAT::Tool::$name";
   return $self->{tools}{$name} ||= $tool_class->new( $self, %args );
}

=head2 $node = $pmat->inref_graph( $sv, %opts )

Traces the tree of inrefs from C<$sv> back towards the known roots, returning
a L<Devel::MAT::Graph> node object representing it, within a graph of reverse
references back to the known roots.

This method will load L<Devel::MAT::Tool::Inrefs> if it isn't yet loaded.

The following named options are recognised:

=over 4

=item depth => INT

If specified, stop recursing after the specified count. A depth of 1 will only
include immediately referring SVs, 2 will print the referrers of those, etc.
Nodes with inrefs that were trimmed because of this limit will appear to be
roots with a special name of C<EDEPTH>.

=item strong => BOOL

=item direct => BOOL

Specifies the type of inrefs followed. By default all inrefs are followed.
Passing C<strong> will follow only strong direct inrefs. Passing C<direct>
will follow only direct inrefs.

=back

=cut

sub inref_graph
{
   my $self = shift;
   my ( $sv, %opts ) = @_;

   my $graph = $opts{graph} //= Devel::MAT::Graph->new( $self->dumpfile );

   $self->load_tool( "Inrefs" );

   if( $sv->immortal ) {
      my $desc = $sv->type eq "UNDEF" ? "undef" :
                 $sv->uv              ? "true" :
                                        "false";
      $graph->add_root( $sv, $desc );
      return $graph->get_sv_node( $sv );
   }

   my $svaddr = $sv->addr;

   foreach ( pairs $self->dumpfile->roots ) {
      my ( $name, $root ) = @$_;
      $root and $svaddr == $root->addr and
         $graph->add_root( $sv, $name ), return $graph->get_sv_node( $sv );
   }

   $graph->add_sv( $sv );

   my @ret = ();
   my @inrefs = $opts{strong} ? $sv->inrefs_strong :
                $opts{direct} ? $sv->inrefs_direct :
                                $sv->inrefs;
   foreach my $ref ( sort_by { $_->name } @inrefs ) {
      if( !defined $ref->sv ) {
         # e.g. "a value on the stack"
         $graph->add_root( $sv, $ref->name );
         push @ret, $ref->name;
         next;
      }

      if( defined $opts{depth} and not $opts{depth} ) {
         $graph->add_root( $sv, "EDEPTH" );
         last;
      }

      my @me;
      if( $graph->has_sv( $ref->sv ) ) {
         $graph->add_ref( $ref->sv, $sv, $ref );
         # Don't recurse into it as it was already found
      }
      else {
         $graph->add_sv( $ref->sv ); # add first to stop inf. loops

         defined $opts{depth} ? $self->inref_graph( $ref->sv, %opts, depth => $opts{depth}-1 )
                              : $self->inref_graph( $ref->sv, %opts );
         $graph->add_ref( $ref->sv, $sv, $ref );
      }
   }

   return $graph->get_sv_node( $sv );
}

=head2 $sv = $pmat->find_symbol( $name )

Attempts to walk the symbol table looking for a symbol of the given name,
which must include the sigil.

 $Package::Name::symbol_name => to return a SCALAR SV
 @Package::Name::symbol_name => to return an ARRAY SV
 %Package::Name::symbol_name => to return a HASH SV
 &Package::Name::symbol_name => to return a CODE SV

=cut

sub find_symbol
{
   my $self = shift;
   my ( $name ) = @_;

   my ( $sigil, $globname ) = $name =~ m/^([\$\@%&])(.*)$/ or
      croak "Could not parse sigil from $name";

   my $glob = $self->find_glob( $globname );

   my $slot = ( $sigil eq '$' ) ? "scalar" :
              ( $sigil eq '@' ) ? "array"  :
              ( $sigil eq '%' ) ? "hash"   :
              ( $sigil eq '&' ) ? "code"   :
                                  die "ARGH"; # won't happen

   my $sv = $glob->$slot or
      croak "\*$globname has no $slot slot";
   return $sv;
}

=head2 $gv = $pmat->find_glob( $name )

Attempts to walk to the symbol table looking for a symbol of the given name,
returning the C<GLOB> object if found.

=cut

sub find_glob
{
   my $self = shift;
   my ( $name ) = @_;

   my ( $parent, $shortname ) = $name =~ m/^(?:(.*)::)?(.+?)$/;

   my $stash;
   if( defined $parent and length $parent ) {
      my $parentgv = $self->find_glob( $parent . "::" );
      $stash = $parentgv->hash or croak "$parent has no hash";
   }
   else {
      $stash = $self->dumpfile->defstash;
   }

   my $gv = $stash->value( $shortname ) or
      croak $stash->stashname . " has no symbol $shortname";
   return $gv;
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
